Index: app/code/Magento/MageConf/view/frontend/ui_component/product_listing.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/code/Magento/MageConf/view/frontend/ui_component/product_listing.xml	(revision )
+++ app/code/Magento/MageConf/view/frontend/ui_component/product_listing.xml	(revision )
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
+<listing xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
+    <columns name="product_columns" class="Magento\Catalog\Ui\Component\Listing\Columns">
+        <column name="name" sortOrder="30" component="Magento_MageConf/js/name_column">
+            <settings>
+                <addField>true</addField>
+                <filter>text</filter>
+                <label translate="true">Name</label>
+            </settings>
+        </column>
+    </columns>
+</listing>
Index: app/code/Magento/MageConf/view/adminhtml/web/js/name_column.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/code/Magento/MageConf/view/adminhtml/web/js/name_column.js	(revision )
+++ app/code/Magento/MageConf/view/adminhtml/web/js/name_column.js	(revision )
@@ -0,0 +1,30 @@
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+/**
+ * @api
+ */
+define([
+    'underscore',
+    'Magento_Ui/js/grid/columns/column',
+    'mageUtils'
+], function (_, Element, utils) {
+    'use strict';
+
+    return Element.extend({
+        defaults: {
+
+        },
+
+        exportSorting: function () {
+            var data = this.source().data,
+                items = data.items;
+
+            items = items.sort(function(a, b) {return this.sorting === 'asc' ? a.name < b.name : a.name > b.name; }.bind(this));
+
+            this.source().date = items;
+        },
+    });
+});
Index: app/code/Magento/Ui/Config/Data.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/code/Magento/Ui/Config/Data.php	(date 1512594713000)
+++ app/code/Magento/Ui/Config/Data.php	(revision )
@@ -93,7 +93,7 @@
      */
     private function initData()
     {
-        $data = $this->cache->load($this->cacheId);
+        $data = false; //$this->cache->load($this->cacheId);
         if (false === $data) {
             /** @var Reader $reader */
             $reader = $this->readerFactory->create(
